### DISCUSSION TITLE
Recursive/nested/transitive queries

### DISCUSSION THREAD

**Discussion by nitnelave** [SeaQL/sea-orm#1498] [General]
## Motivation

I'm currently migrating to sea-orm (from raw sea-query) and I'm looking for a way to express recursive/transitive queries. For instance, I have users who can be members of groups, and groups can inherit from other groups. How do I express the transitive list of the user's groups?
In sea-query, that would be the `.recursive(true)` flag on a `With` clause: https://github.com/SeaQL/sea-query/blob/cfe6bd2ef90fa7483d68364e8563c172de244947/src/query/with.rs#L442

## Proposed Solutions

Still unfamiliar with the API, but the implementation should be relatively straightforward.

**Comment by tyt2y3** [comment]
Welcome! Glad to see you here. If you are migrating from SeaQuery, what you might do is to just call the https://docs.rs/sea-orm/latest/sea_orm/entity/prelude/struct.Select.html#method.query to get a handle to the `SelectStatement` and apply all the SeaQuery expression, and then let SeaORM handle query result.

Reference: https://www.sea-ql.org/SeaORM/docs/advanced-query/custom-select/

@billy1624 it might be worthwhile writing a migration guide from SeaQuery to SeaORM ?

**Comment by billy1624** [comment]
Hey @nitnelave, sorry for the delay. So, you want to select a list of user's groups but the query requires `WITH` clause because it contains recursive queries.

You can get a mutable reference of the SeaQuery select statement from `sea_orm::Select` by [`QuerySelect::query`](https://docs.rs/sea-orm/latest/sea_orm/query/trait.QuerySelect.html#tymethod.query) method. With that I believe you can construct such complex query.

**Comment by nitnelave** [comment]
Oh, I still have to build the query manually? I was hoping that the ORM would take care of that, with a transitive relationship.

Basically, once you define the self relationship between groups (the inheritance) it would be nice to express the transitive closure of that. Then you can query the user's groups, and their transitive closure. WDYT ?

Otherwise, translating queries with conditions on the groups can be tricky, I was hoping to avoid that :)

**Comment by tyt2y3** [comment]
Umm... can you give an example for the "transitive relationship" and the query you are trying to build?

**Comment by nitnelave** [comment]
Sorry for the syntax, I haven't dived into Sea-ORM yet.

Basically, you have a `Users` table with a `UserId`, a `Groups` table with a `GroupId`, a `Membership` table with a `UserId` -> `GroupId` mapping, and a `GroupInheritance` with a `GroupId` -> `GroupId` mapping. I want to get all the groups that the user belongs to, including group inheritance. The query would be something like (WARNING! Approximate SQL):

```sql
WITH resolved_groups AS (
    SELECT group_id
    FROM group_inheritance
    UNION ALL
      SELECT g.group_id
      FROM group_inheritance AS g
      INNER JOIN resolved_groups
      ON g.parent = resolved_groups.group_id
)
SELECT users.user_id, resolved_groups.group_id
FROM users
JOIN memberships ON users.user_id = memberships.user_id
JOIN resolved_groups ON memberships.group_id = resolved_groups.group_id
WHERE users.user_id = "bob"
```

Something like that.

**Reply by hamedai** [reply]
Did you get the answer?

**Reply by nitnelave** [reply]
Nope, sorry. I haven't gotten around to implementing it yet, so I haven't seriously looked into how to express that or whether I'd have to use raw SQL

**Reply by hamedai** [reply]
I need a recursive subquery in Postgres and could not find any solutions but manually load values recursively in a separate raw select statement and use the result values in an orm-based query with "in" statement. :|

**Comment by billy1624** [comment]
> @billy1624 it might be worthwhile writing a migration guide from SeaQuery to SeaORM ?

https://github.com/SeaQL/seaql.github.io/issues/72

**Comment by tyt2y3** [comment]
Given the abstraction we have I am certain one can write a generic function over multiple `EntityTrait` that build specific queries.
I also think that 'recursive self-join for tree-like data structures' is a pattern we can extract as a function.
